# Домашнее задание 
# Тема: Действия с мультииндексами в датафреймах

In [1]:
import pandas as pd
import numpy as np

## Задание 1
### Постройте датафрейм заполненный нулями так, как это показано в ячейке ниже. Первые две колонки являются мультииндексом. Для заполнения датафрейма нулями воспользуйтесь функцией zeros из библиотеки numpy

In [2]:
index = [['Группа 1', 'Группа 1', 'Группа 2', 'Группа 2'], 
         ['Альфа', 'Бета', 'Альфа', 'Бета']]

columns = [['Владивосток', 'Владивосток', 'Казань', 'Новосибирск', 'Новосибирск'],
           ['Кол-во', 'Сумма', 'Сумма', 'Кол-во', 'Сумма']]

df1_result = pd.DataFrame(np.zeros((4,5)), index = index, columns = columns)

In [3]:
df1_result

Владивосток       Казань Новосибирск      
                    Кол-во Сумма  Сумма      Кол-во Сумма
Группа 1 Альфа         0.0   0.0    0.0         0.0   0.0
         Бета          0.0   0.0    0.0         0.0   0.0
Группа 2 Альфа         0.0   0.0    0.0         0.0   0.0
         Бета          0.0   0.0    0.0         0.0   0.0

### Получившийся датафрейм сохраните в csv файл и загрузите на степик

In [7]:
df1_result.to_csv('df1_result.csv')

In [8]:
pd.read_csv('df1_result.csv')

Unnamed: 0 Unnamed: 1 Владивосток Владивосток.1 Казань Новосибирск  \
0        NaN        NaN      Кол-во         Сумма  Сумма      Кол-во   
1   Группа 1      Альфа         0.0           0.0    0.0         0.0   
2   Группа 1       Бета         0.0           0.0    0.0         0.0   
3   Группа 2      Альфа         0.0           0.0    0.0         0.0   
4   Группа 2       Бета         0.0           0.0    0.0         0.0   

  Новосибирск.1  
0         Сумма  
1           0.0  
2           0.0  
3           0.0  
4           0.0

## Задание 2
### Загрузите датафрейм из файла task5.csv. Поменяйте местами уровни так чтобы сначала шел тип парковки (платная/бесплатная), затем округ и район. Отсортируйте так чтобы сначала шли бесплатные парковки в алфавитном порядке по округу, а внутри округа в алфавитном порядке по району*. 

#### *Некоторые названия районов записаны наоборот: вместо район Нижегородский записано Нижегородский район. Из-за этого сортировка по алфавиту будет идти не совсем так, как вы ожидали. Оставим результат сортировки по алфавиту как есть (переносить слово район для корректной сортировки не будем).

In [14]:
df2_task = pd.read_csv('task5.csv', index_col=[0, 1, 2], header=[0, 1]) # не редактируйте ячейку

In [15]:
df2_task

Capacity  \
                                                                           count   
Восточный административный округ Нижегородский район         Бесплатная        1   
                                 район Богородское           Бесплатная        2   
                                                             Платная           1   
                                 район Вешняки               Бесплатная        5   
                                                             Платная           3   
...                                                                          ...   
Южный административный округ     район Чертаново Северное    Платная           2   
                                 район Чертаново Центральное Бесплатная        7   
                                                             Платная           3   
                                 район Чертаново Южное       Бесплатная        9   
                                                             Платная           4   

                                                                             \
                                                                        sum   
Восточный административный округ Нижегородский район         Бесплатная   6   
                                 район Богородское           Бесплатная  14   
                                                             Платная      8   
                                 район Вешняки               Бесплатная  35   
                                                             Платная     17   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная     20   
                                 район Чертаново Центральное Бесплатная  68   
                                                             Платная     21   
                                 район Чертаново Южное       Бесплатная  86   
                                                             Платная     36   

                                                                        rating  \
                                                                           min   
Восточный административный округ Нижегородский район         Бесплатная      8   
                                 район Богородское           Бесплатная      8   
                                                             Платная         0   
                                 район Вешняки               Бесплатная      2   
                                                             Платная         0   
...                                                                        ...   
Южный административный округ     район Чертаново Северное    Платная         2   
                                 район Чертаново Центральное Бесплатная      0   
                                                             Платная         3   
                                 район Чертаново Южное       Бесплатная      3   
                                                             Платная         0   

                                                                             \
                                                                        max   
Восточный административный округ Нижегородский район         Бесплатная   8   
                                 район Богородское           Бесплатная  10   
                                                             Платная      0   
                                 район Вешняки               Бесплатная  10   
                                                             Платная      9   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная      6   
                                 район Чертаново Центральное Бесплатная   9   
                                                             Платная      9   
               

In [16]:
# перестановка индексов

df2_task.swaplevel(1, 2).swaplevel(1, 0)


Capacity  \
                                                                           count   
Бесплатная Восточный административный округ Нижегородский район                1   
                                            район Богородское                  2   
Платная    Восточный административный округ район Богородское                  1   
Бесплатная Восточный административный округ район Вешняки                      5   
Платная    Восточный административный округ район Вешняки                      3   
...                                                                          ...   
           Южный административный округ     район Чертаново Северное           2   
Бесплатная Южный административный округ     район Чертаново Центральное        7   
Платная    Южный административный округ     район Чертаново Центральное        3   
Бесплатная Южный административный округ     район Чертаново Южное              9   
Платная    Южный административный округ     район Чертаново Южное              4   

                                                                             \
                                                                        sum   
Бесплатная Восточный административный округ Нижегородский район           6   
                                            район Богородское            14   
Платная    Восточный административный округ район Богородское             8   
Бесплатная Восточный административный округ район Вешняки                35   
Платная    Восточный административный округ район Вешняки                17   
...                                                                      ..   
           Южный административный округ     район Чертаново Северное     20   
Бесплатная Южный административный округ     район Чертаново Центральное  68   
Платная    Южный административный округ     район Чертаново Центральное  21   
Бесплатная Южный административный округ     район Чертаново Южное        86   
Платная    Южный административный округ     район Чертаново Южное        36   

                                                                        rating  \
                                                                           min   
Бесплатная Восточный административный округ Нижегородский район              8   
                                            район Богородское                8   
Платная    Восточный административный округ район Богородское                0   
Бесплатная Восточный административный округ район Вешняки                    2   
Платная    Восточный административный округ район Вешняки                    0   
...                                                                        ...   
           Южный административный округ     район Чертаново Северное         2   
Бесплатная Южный административный округ     район Чертаново Центральное      0   
Платная    Южный административный округ     район Чертаново Центральное      3   
Бесплатная Южный административный округ     район Чертаново Южное            3   
Платная    Южный административный округ     район Чертаново Южное            0   

                                                                             \
                                                                        max   
Бесплатная Восточный административный округ Нижегородский район           8   
                                            район Богородское            10   
Платная    Восточный административный округ район Богородское             0   
Бесплатная Восточный административный округ район Вешняки                10   
Платная    Восточный административный округ район Вешняки                 9   
...                                                                      ..   
           Южный административный округ     район Чертаново Северное      6   
Бесплатная Южный административный округ     район Чертаново Центральное   9   
Платная    Южный административный округ     район Чертаново Центральное   9   
Бесплатная Южны

#### Обратите внимание, наш датафрейм немного усложнился: у мультииндекса строк пропали названия для уровней (добавлять их не стоит) и появился мультииндекс для колонок. Теперь в датафрейме содержится дополнительная информация о рейтинге парковки: маинимальное, максимальное и среднее значение

In [17]:
df2_result = df2_task.swaplevel().swaplevel(1,0)

In [18]:
df2_result

Capacity  \
                                                                           count   
Бесплатная Восточный административный округ Нижегородский район                1   
                                            район Богородское                  2   
Платная    Восточный административный округ район Богородское                  1   
Бесплатная Восточный административный округ район Вешняки                      5   
Платная    Восточный административный округ район Вешняки                      3   
...                                                                          ...   
           Южный административный округ     район Чертаново Северное           2   
Бесплатная Южный административный округ     район Чертаново Центральное        7   
Платная    Южный административный округ     район Чертаново Центральное        3   
Бесплатная Южный административный округ     район Чертаново Южное              9   
Платная    Южный административный округ     район Чертаново Южное              4   

                                                                             \
                                                                        sum   
Бесплатная Восточный административный округ Нижегородский район           6   
                                            район Богородское            14   
Платная    Восточный административный округ район Богородское             8   
Бесплатная Восточный административный округ район Вешняки                35   
Платная    Восточный административный округ район Вешняки                17   
...                                                                      ..   
           Южный административный округ     район Чертаново Северное     20   
Бесплатная Южный административный округ     район Чертаново Центральное  68   
Платная    Южный административный округ     район Чертаново Центральное  21   
Бесплатная Южный административный округ     район Чертаново Южное        86   
Платная    Южный административный округ     район Чертаново Южное        36   

                                                                        rating  \
                                                                           min   
Бесплатная Восточный административный округ Нижегородский район              8   
                                            район Богородское                8   
Платная    Восточный административный округ район Богородское                0   
Бесплатная Восточный административный округ район Вешняки                    2   
Платная    Восточный административный округ район Вешняки                    0   
...                                                                        ...   
           Южный административный округ     район Чертаново Северное         2   
Бесплатная Южный административный округ     район Чертаново Центральное      0   
Платная    Южный административный округ     район Чертаново Центральное      3   
Бесплатная Южный административный округ     район Чертаново Южное            3   
Платная    Южный административный округ     район Чертаново Южное            0   

                                                                             \
                                                                        max   
Бесплатная Восточный административный округ Нижегородский район           8   
                                            район Богородское            10   
Платная    Восточный административный округ район Богородское             0   
Бесплатная Восточный административный округ район Вешняки                10   
Платная    Восточный административный округ район Вешняки                 9   
...                                                                      ..   
           Южный административный округ     район Чертаново Северное      6   
Бесплатная Южный административный округ     район Чертаново Центральное   9   
Платная    Южный административный округ     район Чертаново Центральное   9   
Бесплатная Южны

In [41]:
# сортировка:

# Отсортируйте так чтобы сначала шли бесплатные парковки в алфавитном порядке по округу, а внутри 
# округа в алфавитном порядке по району*.

# *Некоторые названия районов записаны наоборот: вместо район Нижегородский записано Нижегородский район.
#     Из-за этого сортировка по алфавиту будет идти не совсем так, как вы ожидали. Оставим результат 
#     сортировки по алфавиту как есть (переносить слово район для корректной сортировки не будем).



# можно в sort_index указывать кортеж (level=(0,1,2))



df2_result.sort_index(level = 0)

Capacity  \
                                                                            count   
Бесплатная Восточный административный округ Нижегородский район                 1   
                                            район Богородское                   2   
                                            район Вешняки                       5   
                                            район Восточное Измайлово           1   
                                            район Гольяново                     2   
...                                                                           ...   
Платная    Южный административный округ     район Орехово-Борисово Южное       10   
                                            район Царицыно                      4   
                                            район Чертаново Северное            2   
                                            район Чертаново Центральное         3   
                                            район Чертаново Южное               4   

                                                                              \
                                                                         sum   
Бесплатная Восточный административный округ Нижегородский район            6   
                                            район Богородское             14   
                                            район Вешняки                 35   
                                            район Восточное Измайлово      8   
                                            район Гольяново                9   
...                                                                       ..   
Платная    Южный административный округ     район Орехово-Борисово Южное  85   
                                            район Царицыно                38   
                                            район Чертаново Северное      20   
                                            район Чертаново Центральное   21   
                                            район Чертаново Южное         36   

                                                                         rating  \
                                                                            min   
Бесплатная Восточный административный округ Нижегородский район               8   
                                            район Богородское                 8   
                                            район Вешняки                     2   
                                            район Восточное Измайлово         6   
                                            район Гольяново                   0   
...                                                                         ...   
Платная    Южный административный округ     район Орехово-Борисово Южное      0   
                                            район Царицыно                    2   
                                            район Чертаново Северное          2   
                                            район Чертаново Центральное       3   
                                            район Чертаново Южное             0   

                                                                              \
                                                                         max   
Бесплатная Восточный административный округ Нижегородский район            8   
                                            район Богородское             10   
                                            район Вешняки                 10   
                                            район Восточное Измайлово      6   
                                            район Гольяново               10   
...                                                                       ..   
Платная    Южный административный округ     район Орехово-Борисово Южное   7   
                                            район Царицыно                10   
                                            ра

In [42]:
df2_result = df2_result.sort_index(level = 0)

In [44]:
df2_result.head(27)

Capacity  \
                                                                          count   
Бесплатная Восточный административный округ Нижегородский район               1   
                                            район Богородское                 2   
                                            район Вешняки                     5   
                                            район Восточное Измайлово         1   
                                            район Гольяново                   2   
                                            район Ивановское                  2   
                                            район Измайлово                   2   
                                            район Косино-Ухтомский            3   
                                            район Метрогородок                4   
                                            район Новогиреево                 2   
                                            район Новокосино                  1   
                                            район Перово                      5   
                                            район Преображенское              3   
                                            район Северное Измайлово          1   
                                            район Соколиная Гора              4   
                                            район Сокольники                  4   
           Западный административный округ  Можайский район                   1   
                                            район Дорогомилово                1   
                                            район Крылатское                  2   
                                            район Кунцево                     1   
                                            район Ново-Переделкино            4   
                                            район Очаково-Матвеевское         1   
                                            район Проспект Вернадского        2   
                                            район Раменки                     4   
                                            район Солнцево                    2   
                                            район Тропарёво-Никулино          1   
                                            район Филёвский Парк              1   

                                                                            \
                                                                       sum   
Бесплатная Восточный административный округ Нижегородский район          6   
                                            район Богородское           14   
                                            район Вешняки               35   
                                            район Восточное Измайлово    8   
                                            район Гольяново              9   
                                            район Ивановское            16   
                                            район Измайлово             11   
                                            район Косино-Ухтомский      18   
                                            район Метрогородок          18   
                                            район Новогиреево           11   
                                            район Новокосино             1   
                                            район Перово                35   
                                            район Преображенское        21   
                                            район Северное Измайлово     1   
                                            район Соколиная Гора        16   
                                            район Сокольники            28   
           Западный административный округ  Можайский район              8   
                                            район Дорогомилово           8   
                                            район Крылатское            18   
                         

In [45]:
df2_result = df2_result.head(27)

### Сохраните на степик только первые 27 записей датафрейма.

In [46]:
df2_result.to_csv('df2_result.csv')

## Задание 3
### Загрузите датафрейм из файла task5.csv. Из мультииндекса строк сбросьте уровень в котором содержится информация о типе парковки. Сохраните данные из этого уровня в отдельную колонку под названием `type`. Из целого датафрейма вырезаем только данные о рейтинге, сортируем записи по среднему рейтингу в порядке убывания и загружаем на степик ТОП-10 записей.

In [84]:
df3_task = pd.read_csv('task5.csv', index_col=[0, 1, 2], header=[0, 1]) # не редактируйте ячейку

In [85]:
df3_task

Capacity  \
                                                                           count   
Восточный административный округ Нижегородский район         Бесплатная        1   
                                 район Богородское           Бесплатная        2   
                                                             Платная           1   
                                 район Вешняки               Бесплатная        5   
                                                             Платная           3   
...                                                                          ...   
Южный административный округ     район Чертаново Северное    Платная           2   
                                 район Чертаново Центральное Бесплатная        7   
                                                             Платная           3   
                                 район Чертаново Южное       Бесплатная        9   
                                                             Платная           4   

                                                                             \
                                                                        sum   
Восточный административный округ Нижегородский район         Бесплатная   6   
                                 район Богородское           Бесплатная  14   
                                                             Платная      8   
                                 район Вешняки               Бесплатная  35   
                                                             Платная     17   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная     20   
                                 район Чертаново Центральное Бесплатная  68   
                                                             Платная     21   
                                 район Чертаново Южное       Бесплатная  86   
                                                             Платная     36   

                                                                        rating  \
                                                                           min   
Восточный административный округ Нижегородский район         Бесплатная      8   
                                 район Богородское           Бесплатная      8   
                                                             Платная         0   
                                 район Вешняки               Бесплатная      2   
                                                             Платная         0   
...                                                                        ...   
Южный административный округ     район Чертаново Северное    Платная         2   
                                 район Чертаново Центральное Бесплатная      0   
                                                             Платная         3   
                                 район Чертаново Южное       Бесплатная      3   
                                                             Платная         0   

                                                                             \
                                                                        max   
Восточный административный округ Нижегородский район         Бесплатная   8   
                                 район Богородское           Бесплатная  10   
                                                             Платная      0   
                                 район Вешняки               Бесплатная  10   
                                                             Платная      9   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная      6   
                                 район Чертаново Центральное Бесплатная   9   
                                                             Платная      9   
               

In [86]:
df3_task.index.names

FrozenList([None, None, None])

In [74]:
# оставляем только rating, сортируем по mean на убывание

df3_task['rating'].sort_values(by = 'mean', ascending = False)

min  \
Юго-Восточный административный округ    район Некрасовка         Бесплатная   10   
Зеленоградский административный округ   район Силино             Платная      10   
Северный административный округ         район Сокол              Платная      10   
Западный административный округ         район Марьино            Платная      10   
                                        район Филёвский Парк     Бесплатная   10   
...                                                                          ...   
Восточный административный округ        район Северное Измайлово Бесплатная    1   
Северо-Восточный административный округ район Марфино            Платная       1   
Юго-Западный административный округ     Обручевский район        Платная       0   
Северо-Восточный административный округ район Свиблово           Платная       0   
Восточный административный округ        район Богородское        Платная       0   

                                                                             max  \
Юго-Восточный административный округ    район Некрасовка         Бесплатная   10   
Зеленоградский административный округ   район Силино             Платная      10   
Северный административный округ         район Сокол              Платная      10   
Западный административный округ         район Марьино            Платная      10   
                                        район Филёвский Парк     Бесплатная   10   
...                                                                          ...   
Восточный административный округ        район Северное Измайлово Бесплатная    1   
Северо-Восточный административный округ район Марфино            Платная       1   
Юго-Западный административный округ     Обручевский район        Платная       1   
Северо-Восточный административный округ район Свиблово           Платная       0   
Восточный административный округ        район Богородское        Платная       0   

                                                                              mean  
Юго-Восточный административный округ    район Некрасовка         Бесплатная  10.00  
Зеленоградский административный округ   район Силино             Платная     10.00  
Северный административный округ         район Сокол              Платная     10.00  
Западный административный округ         район Марьино            Платная     10.00  
                                        район Филёвский Парк     Бесплатная  10.00  
...                                                                            ...  
Восточный административный округ        район Северное Измайлово Бесплатная   1.00  
Северо-Восточный административный округ район Марфино            Платная      1.00  
Юго-Западный административный округ     Обручевский район        Платная      0.25  
Северо-Восточный административный округ район Свиблово           Платная      0.00  
Восточный административный округ        район Богородское        Платная      0.00  

[247 rows x 3 columns]

In [75]:
# плюс сбрасываем индексы

df3_task['rating'].sort_values(by = 'mean', ascending = False).reset_index()

level_0                   level_1  \
0       Юго-Восточный административный округ          район Некрасовка   
1      Зеленоградский административный округ              район Силино   
2            Северный административный округ               район Сокол   
3            Западный административный округ             район Марьино   
4            Западный административный округ      район Филёвский Парк   
..                                       ...                       ...   
242         Восточный административный округ  район Северное Измайлово   
243  Северо-Восточный административный округ             район Марфино   
244      Юго-Западный административный округ         Обручевский район   
245  Северо-Восточный административный округ            район Свиблово   
246         Восточный административный округ         район Богородское   

        level_2  min  max   mean  
0    Бесплатная   10   10  10.00  
1       Платная   10   10  10.00  
2       Платная   10   10  10.00  
3       Платная   10   10  10.00  
4    Бесплатная   10   10  10.00  
..          ...  ...  ...    ...  
242  Бесплатная    1    1   1.00  
243     Платная    1    1   1.00  
244     Платная    0    1   0.25  
245     Платная    0    0   0.00  
246     Платная    0    0   0.00  

[247 rows x 6 columns]

In [76]:
# присваиваем результат переменной r

r = df3_task['rating'].sort_values(by = 'mean', ascending = False).reset_index()

In [77]:
# устанавливаем индексы:

r.set_index(['level_0', 'level_1'])

level_2  \
level_0                                 level_1                                
Юго-Восточный административный округ    район Некрасовка          Бесплатная   
Зеленоградский административный округ   район Силино                 Платная   
Северный административный округ         район Сокол                  Платная   
Западный административный округ         район Марьино                Платная   
                                        район Филёвский Парк      Бесплатная   
...                                                                      ...   
Восточный административный округ        район Северное Измайлово  Бесплатная   
Северо-Восточный административный округ район Марфино                Платная   
Юго-Западный административный округ     Обручевский район            Платная   
Северо-Восточный административный округ район Свиблово               Платная   
Восточный административный округ        район Богородское            Платная   

                                                                  min  max  \
level_0                                 level_1                              
Юго-Восточный административный округ    район Некрасовка           10   10   
Зеленоградский административный округ   район Силино               10   10   
Северный административный округ         район Сокол                10   10   
Западный административный округ         район Марьино              10   10   
                                        район Филёвский Парк       10   10   
...                                                               ...  ...   
Восточный административный округ        район Северное Измайлово    1    1   
Северо-Восточный административный округ район Марфино               1    1   
Юго-Западный административный округ     Обручевский район           0    1   
Северо-Восточный административный округ район Свиблово              0    0   
Восточный административный округ        район Богородское           0    0   

                                                                   mean  
level_0                                 level_1                          
Юго-Восточный административный округ    район Некрасовка          10.00  
Зеленоградский административный округ   район Силино              10.00  
Северный административный округ         район Сокол               10.00  
Западный административный округ         район Марьино             10.00  
                                        район Филёвский Парк      10.00  
...                                                                 ...  
Восточный административный округ        район Северное Измайлово   1.00  
Северо-Восточный административный округ район Марфино              1.00  
Юго-Западный административный округ     Обручевский район          0.25  
Северо-Восточный административный округ район Свиблово             0.00  
Восточный административный округ        район Богородское          0.00  

[247 rows x 4 columns]

In [78]:
rr = r.set_index(['level_0', 'level_1'])

In [79]:
# задаем пустые индексы

rr.index.names = [None, None]

In [80]:
# именуем колонки

rr.columns = ['type', 'min', 'max', 'mean']

In [87]:
df3_result = rr.head(10)

In [88]:
df3_result

type  min  \
Юго-Восточный административный округ  район Некрасовка      Бесплатная   10   
Зеленоградский административный округ район Силино             Платная   10   
Северный административный округ       район Сокол              Платная   10   
Западный административный округ       район Марьино            Платная   10   
                                      район Филёвский Парк  Бесплатная   10   
Юго-Восточный административный округ  район Печатники       Бесплатная    9   
Западный административный округ       Можайский район       Бесплатная    9   
                                      район Дорогомилово    Бесплатная    9   
Юго-Восточный административный округ  Гагаринский район     Бесплатная    9   
Новомосковский административный округ поселение Сосенское      Платная    9   

                                                            max  mean  
Юго-Восточный административный округ  район Некрасовка       10  10.0  
Зеленоградский административный округ район Силино           10  10.0  
Северный административный округ       район Сокол            10  10.0  
Западный административный округ       район Марьино          10  10.0  
                                      район Филёвский Парк   10  10.0  
Юго-Восточный административный округ  район Печатники         9   9.0  
Западный административный округ       Можайский район         9   9.0  
                                      район Дорогомилово      9   9.0  
Юго-Восточный административный округ  Гагаринский район       9   9.0  
Новомосковский административный округ поселение Сосенское     9   9.0

### Сохраните и отправьте файл на степик.

In [122]:
df3_result.to_csv('df3_result.csv')

## Задание 4 
### В одном из прошлых заданий мы выбирали все районы восточного административного округа и искали где больше всего мест на бесплатных парковках. Загрузите датафрейм из файла task5.csv и попробуйте найти в каком округе больше всего парковок (и платных и бесплатных вместе). Учитывайте только те районы, где средний рейтинг парковок больше или равен 5 

In [134]:
df4_task = pd.read_csv('task5.csv', index_col=[0, 1, 2], header=[0, 1]) # не редактируйте ячейку

In [137]:
df4_task

Capacity  \
                                                                           count   
Восточный административный округ Нижегородский район         Бесплатная        1   
                                 район Богородское           Бесплатная        2   
                                                             Платная           1   
                                 район Вешняки               Бесплатная        5   
                                                             Платная           3   
...                                                                          ...   
Южный административный округ     район Чертаново Северное    Платная           2   
                                 район Чертаново Центральное Бесплатная        7   
                                                             Платная           3   
                                 район Чертаново Южное       Бесплатная        9   
                                                             Платная           4   

                                                                             \
                                                                        sum   
Восточный административный округ Нижегородский район         Бесплатная   6   
                                 район Богородское           Бесплатная  14   
                                                             Платная      8   
                                 район Вешняки               Бесплатная  35   
                                                             Платная     17   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная     20   
                                 район Чертаново Центральное Бесплатная  68   
                                                             Платная     21   
                                 район Чертаново Южное       Бесплатная  86   
                                                             Платная     36   

                                                                        rating  \
                                                                           min   
Восточный административный округ Нижегородский район         Бесплатная      8   
                                 район Богородское           Бесплатная      8   
                                                             Платная         0   
                                 район Вешняки               Бесплатная      2   
                                                             Платная         0   
...                                                                        ...   
Южный административный округ     район Чертаново Северное    Платная         2   
                                 район Чертаново Центральное Бесплатная      0   
                                                             Платная         3   
                                 район Чертаново Южное       Бесплатная      3   
                                                             Платная         0   

                                                                             \
                                                                        max   
Восточный административный округ Нижегородский район         Бесплатная   8   
                                 район Богородское           Бесплатная  10   
                                                             Платная      0   
                                 район Вешняки               Бесплатная  10   
                                                             Платная      9   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная      6   
                                 район Чертаново Центральное Бесплатная   9   
                                                             Платная      9   
               

In [139]:
df4_task['Средний рейтинг'] = df4_task['rating']['mean']

In [126]:
df4_task[df4_task['Средний рейтинг']>= 5]

Capacity  \
                                                                           count   
Восточный административный округ Нижегородский район         Бесплатная        1   
                                 район Богородское           Бесплатная        2   
                                 район Вешняки               Бесплатная        5   
                                 район Восточное Измайлово   Бесплатная        1   
                                                             Платная           1   
...                                                                          ...   
Южный административный округ     район Царицыно              Платная           4   
                                 район Чертаново Северное    Бесплатная        6   
                                 район Чертаново Центральное Платная           3   
                                 район Чертаново Южное       Бесплатная        9   
                                                             Платная           4   

                                                                             \
                                                                        sum   
Восточный административный округ Нижегородский район         Бесплатная   6   
                                 район Богородское           Бесплатная  14   
                                 район Вешняки               Бесплатная  35   
                                 район Восточное Измайлово   Бесплатная   8   
                                                             Платная     10   
...                                                                      ..   
Южный административный округ     район Царицыно              Платная     38   
                                 район Чертаново Северное    Бесплатная  58   
                                 район Чертаново Центральное Платная     21   
                                 район Чертаново Южное       Бесплатная  86   
                                                             Платная     36   

                                                                        rating  \
                                                                           min   
Восточный административный округ Нижегородский район         Бесплатная      8   
                                 район Богородское           Бесплатная      8   
                                 район Вешняки               Бесплатная      2   
                                 район Восточное Измайлово   Бесплатная      6   
                                                             Платная         7   
...                                                                        ...   
Южный административный округ     район Царицыно              Платная         2   
                                 район Чертаново Северное    Бесплатная      2   
                                 район Чертаново Центральное Платная         3   
                                 район Чертаново Южное       Бесплатная      3   
                                                             Платная         0   

                                                                             \
                                                                        max   
Восточный административный округ Нижегородский район         Бесплатная   8   
                                 район Богородское           Бесплатная  10   
                                 район Вешняки               Бесплатная  10   
                                 район Восточное Измайлово   Бесплатная   6   
                                                             Платная      7   
...                                                                      ..   
Южный административный округ     район Царицыно              Платная     10   
                                 район Чертаново Северное    Бесплатная  10   
                                 район Чертаново Центральное Платная      9   
               

In [127]:
df4_task[df4_task['Средний рейтинг']>= 5].sum(level = 0)

C:\Users\roman\AppData\Local\Temp\ipykernel_86728\329201025.py:1: FutureWarning: Using the level keyword in DataFrame and Series aggregations is deprecated and will be removed in a future version. Use groupby instead. df.sum(level=1) should use df.groupby(level=1).sum().
  df4_task[df4_task['Средний рейтинг']>= 5].sum(level = 0)


Capacity      rating                   \
                                           count  sum    min  max        mean   
Восточный административный округ              35  216     61  119   92.450000   
Западный административный округ               28  219     82  109   95.500000   
Зеленоградский административный округ         13   94     19   36   28.100000   
Новомосковский административный округ          8   96     20   35   27.500000   
Северный административный округ               32  255     81  124  104.150000   
Северо-Восточный административный округ      110  953     52  183  124.219170   
Северо-Западный административный округ        14   99     18   40   28.866667   
Центральный административный округ            98  305     20  100   63.394824   
Юго-Восточный административный округ          35  248     66  115   91.000000   
Юго-Западный административный округ           27  241     26   66   49.150000   
Южный административный округ                  60  498     70  160  116.800000   

                                        Средний рейтинг  
                                                         
Восточный административный округ              92.450000  
Западный административный округ               95.500000  
Зеленоградский административный округ         28.100000  
Новомосковский административный округ         27.500000  
Северный административный округ              104.150000  
Северо-Восточный административный округ      124.219170  
Северо-Западный административный округ        28.866667  
Центральный административный округ            63.394824  
Юго-Восточный административный округ          91.000000  
Юго-Западный административный округ           49.150000  
Южный административный округ                 116.800000

In [128]:
df4_task[df4_task['Средний рейтинг']>= 5].sum(level = 0)['Capacity']

C:\Users\roman\AppData\Local\Temp\ipykernel_86728\229977934.py:1: FutureWarning: Using the level keyword in DataFrame and Series aggregations is deprecated and will be removed in a future version. Use groupby instead. df.sum(level=1) should use df.groupby(level=1).sum().
  df4_task[df4_task['Средний рейтинг']>= 5].sum(level = 0)['Capacity']


count  sum
Восточный административный округ            35  216
Западный административный округ             28  219
Зеленоградский административный округ       13   94
Новомосковский административный округ        8   96
Северный административный округ             32  255
Северо-Восточный административный округ    110  953
Северо-Западный административный округ      14   99
Центральный административный округ          98  305
Юго-Восточный административный округ        35  248
Юго-Западный административный округ         27  241
Южный административный округ                60  498

In [129]:
df4_task[df4_task['Средний рейтинг']>= 5].sum(level = 0)['Capacity'].sort_values(by = 'count')

C:\Users\roman\AppData\Local\Temp\ipykernel_86728\2794587217.py:1: FutureWarning: Using the level keyword in DataFrame and Series aggregations is deprecated and will be removed in a future version. Use groupby instead. df.sum(level=1) should use df.groupby(level=1).sum().
  df4_task[df4_task['Средний рейтинг']>= 5].sum(level = 0)['Capacity'].sort_values(by = 'count')


count  sum
Новомосковский административный округ        8   96
Зеленоградский административный округ       13   94
Северо-Западный административный округ      14   99
Юго-Западный административный округ         27  241
Западный административный округ             28  219
Северный административный округ             32  255
Восточный административный округ            35  216
Юго-Восточный административный округ        35  248
Южный административный округ                60  498
Центральный административный округ          98  305
Северо-Восточный административный округ    110  953

### Правильный ответ выберите на степике

## Задание 5
### Загрузите датафрейм из файла task5.csv. В этой задаче вы попробуете перейти от мультииндекса к обычному индексу с сохранением информации об уровнях. Сейчас на нулевом уровне колонки разбиваются на Capacity и rating, а на первом уровне Capacity и rating детализируются на count/sum и min/max/mean соответственно.
### Попытайтесь уйти от иерархии с сохранением информации о ней: сделайте обычный индекс с названием колонок:
#### Capacity_count, Capacity_sum, rating_min, rating_max, rating_mean

#### Не пытайтесь руками составить список и присвоить его т.к. на практике у вас может быть много колонок и руками перебирать будет долго. А еще можно что-то перепутать и тогда будет ошибка в последующих расчётах. Попробуйте перебрать значения мультииндекса в цикле и собрать тот самый список, который уже и присвоите обратно в датафрейм

In [140]:
df5_task = pd.read_csv('task5.csv', index_col=[0, 1, 2], header=[0, 1]) # не редактируйте ячейку

In [145]:
df5_task

Capacity  \
                                                                           count   
Восточный административный округ Нижегородский район         Бесплатная        1   
                                 район Богородское           Бесплатная        2   
                                                             Платная           1   
                                 район Вешняки               Бесплатная        5   
                                                             Платная           3   
...                                                                          ...   
Южный административный округ     район Чертаново Северное    Платная           2   
                                 район Чертаново Центральное Бесплатная        7   
                                                             Платная           3   
                                 район Чертаново Южное       Бесплатная        9   
                                                             Платная           4   

                                                                             \
                                                                        sum   
Восточный административный округ Нижегородский район         Бесплатная   6   
                                 район Богородское           Бесплатная  14   
                                                             Платная      8   
                                 район Вешняки               Бесплатная  35   
                                                             Платная     17   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная     20   
                                 район Чертаново Центральное Бесплатная  68   
                                                             Платная     21   
                                 район Чертаново Южное       Бесплатная  86   
                                                             Платная     36   

                                                                        rating  \
                                                                           min   
Восточный административный округ Нижегородский район         Бесплатная      8   
                                 район Богородское           Бесплатная      8   
                                                             Платная         0   
                                 район Вешняки               Бесплатная      2   
                                                             Платная         0   
...                                                                        ...   
Южный административный округ     район Чертаново Северное    Платная         2   
                                 район Чертаново Центральное Бесплатная      0   
                                                             Платная         3   
                                 район Чертаново Южное       Бесплатная      3   
                                                             Платная         0   

                                                                             \
                                                                        max   
Восточный административный округ Нижегородский район         Бесплатная   8   
                                 район Богородское           Бесплатная  10   
                                                             Платная      0   
                                 район Вешняки               Бесплатная  10   
                                                             Платная      9   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная      6   
                                 район Чертаново Центральное Бесплатная   9   
                                                             Платная      9   
               

In [142]:
df5_task.columns

MultiIndex([('Capacity', 'count'),
            ('Capacity',   'sum'),
            (  'rating',   'min'),
            (  'rating',   'max'),
            (  'rating',  'mean')],
           )

In [147]:
list_columns_new = []
for i in df5_task.columns:
    print (i[0] + '_' + i[1])
    list_columns_new.append(i[0] + '_' + i[1])

Capacity_count
Capacity_sum
rating_min
rating_max
rating_mean


In [148]:
list_columns_new

['Capacity_count', 'Capacity_sum', 'rating_min', 'rating_max', 'rating_mean']

In [160]:
df5_task.columns = list_columns_new

In [161]:
df5_task

Capacity_count  \
Восточный административный округ Нижегородский район         Бесплатная               1   
                                 район Богородское           Бесплатная               2   
                                                             Платная                  1   
                                 район Вешняки               Бесплатная               5   
                                                             Платная                  3   
...                                                                                 ...   
Южный административный округ     район Чертаново Северное    Платная                  2   
                                 район Чертаново Центральное Бесплатная               7   
                                                             Платная                  3   
                                 район Чертаново Южное       Бесплатная               9   
                                                             Платная                  4   

                                                                         Capacity_sum  \
Восточный административный округ Нижегородский район         Бесплатная             6   
                                 район Богородское           Бесплатная            14   
                                                             Платная                8   
                                 район Вешняки               Бесплатная            35   
                                                             Платная               17   
...                                                                               ...   
Южный административный округ     район Чертаново Северное    Платная               20   
                                 район Чертаново Центральное Бесплатная            68   
                                                             Платная               21   
                                 район Чертаново Южное       Бесплатная            86   
                                                             Платная               36   

                                                                         rating_min  \
Восточный административный округ Нижегородский район         Бесплатная           8   
                                 район Богородское           Бесплатная           8   
                                                             Платная              0   
                                 район Вешняки               Бесплатная           2   
                                                             Платная              0   
...                                                                             ...   
Южный административный округ     район Чертаново Северное    Платная              2   
                                 район Чертаново Центральное Бесплатная           0   
                                                             Платная              3   
                                 район Чертаново Южное       Бесплатная           3   
                                                             Платная              0   

                                                                         rating_max  \
Восточный административный округ Нижегородский район         Бесплатная           8   
                                 район Богородское           Бесплатная          10   
                                                             Платная              0   
                                 район Вешняки               Бесплатная          10   
                                                             Платная              9   
...                                                                             ...   
Южный административный округ     район Чертаново Северное    Платная              6   
                                 район Чертаново Центральное Бесплатная           9   
                                                             Платная              9

In [162]:
df5_result = df5_task

### Сохраните и отправьте файл на степик.

In [163]:
df5_result.to_csv('df5_result.csv')

## Задача 6
### Загрузите данные из task5.csv. По всей Москве рассчитайте:
### -Общее количество бесплатных парковок,
### -Общее количество платных парковок,
### -Общее количество мест на бесплатных парковках,
### -Общее количество мест на платных парковках

### Введите в поле с ответом получившиеся значения через пробел. В ответе должны быть целые числа
### Пример: 10 12 90 110

In [183]:
df6_task = pd.read_csv('task5.csv', index_col=[0, 1, 2], header=[0, 1]) 

In [184]:
df6_task

Capacity  \
                                                                           count   
Восточный административный округ Нижегородский район         Бесплатная        1   
                                 район Богородское           Бесплатная        2   
                                                             Платная           1   
                                 район Вешняки               Бесплатная        5   
                                                             Платная           3   
...                                                                          ...   
Южный административный округ     район Чертаново Северное    Платная           2   
                                 район Чертаново Центральное Бесплатная        7   
                                                             Платная           3   
                                 район Чертаново Южное       Бесплатная        9   
                                                             Платная           4   

                                                                             \
                                                                        sum   
Восточный административный округ Нижегородский район         Бесплатная   6   
                                 район Богородское           Бесплатная  14   
                                                             Платная      8   
                                 район Вешняки               Бесплатная  35   
                                                             Платная     17   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная     20   
                                 район Чертаново Центральное Бесплатная  68   
                                                             Платная     21   
                                 район Чертаново Южное       Бесплатная  86   
                                                             Платная     36   

                                                                        rating  \
                                                                           min   
Восточный административный округ Нижегородский район         Бесплатная      8   
                                 район Богородское           Бесплатная      8   
                                                             Платная         0   
                                 район Вешняки               Бесплатная      2   
                                                             Платная         0   
...                                                                        ...   
Южный административный округ     район Чертаново Северное    Платная         2   
                                 район Чертаново Центральное Бесплатная      0   
                                                             Платная         3   
                                 район Чертаново Южное       Бесплатная      3   
                                                             Платная         0   

                                                                             \
                                                                        max   
Восточный административный округ Нижегородский район         Бесплатная   8   
                                 район Богородское           Бесплатная  10   
                                                             Платная      0   
                                 район Вешняки               Бесплатная  10   
                                                             Платная      9   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная      6   
                                 район Чертаново Центральное Бесплатная   9   
                                                             Платная      9   
               

In [167]:
# Общее количество парковок

df6_task['Capacity']['count'].sum()

1047

In [172]:
# Общее количество бесплатных парковок

df6_task.reset_index()

level_0                      level_1  \
                                                                     
0    Восточный административный округ          Нижегородский район   
1    Восточный административный округ            район Богородское   
2    Восточный административный округ            район Богородское   
3    Восточный административный округ                район Вешняки   
4    Восточный административный округ                район Вешняки   
..                                ...                          ...   
242      Южный административный округ     район Чертаново Северное   
243      Южный административный округ  район Чертаново Центральное   
244      Южный административный округ  район Чертаново Центральное   
245      Южный административный округ        район Чертаново Южное   
246      Южный административный округ        район Чертаново Южное   

        level_2 Capacity     rating                
                   count sum    min max      mean  
0    Бесплатная        1   6      8   8  8.000000  
1    Бесплатная        2  14      8  10  9.000000  
2       Платная        1   8      0   0  0.000000  
3    Бесплатная        5  35      2  10  6.800000  
4       Платная        3  17      0   9  4.666667  
..          ...      ...  ..    ...  ..       ...  
242     Платная        2  20      2   6  4.000000  
243  Бесплатная        7  68      0   9  3.571429  
244     Платная        3  21      3   9  6.000000  
245  Бесплатная        9  86      3  10  6.666667  
246     Платная        4  36      0   9  5.000000  

[247 rows x 8 columns]

In [180]:
df6_task.reset_index()[df6_task.reset_index()['level_2'] == 'Бесплатная']['Capacity']['count'].sum()

518

In [185]:
# # Общее количество платных парковок

df6_task.reset_index()[df6_task.reset_index()['level_2'] == 'Платная']['Capacity']['count'].sum()

529

In [181]:
# Общее количество мест на бесплатных парковках

df6_task.reset_index()[df6_task.reset_index()['level_2'] == 'Бесплатная']['Capacity']['sum'].sum()


3455

In [182]:
# Общее количество мест на платных парковках

df6_task.reset_index()[df6_task.reset_index()['level_2'] == 'Платная']['Capacity']['sum'].sum()

3642